### ISSUE TITLE
Productionization planning

### ISSUE THREAD

**Issue by cczhu** [CityofToronto/bdit_triplinker#1]
Overview of the productionization procedure, including a brainstorm of possible additional investigations or features to implement.  Time estimates assume no obstacles are encountered (so multiplying by 1.5 - 3 may be reasonable).

Essential:
- Port `triplinker` code from `bdit_vfh` to this repo.  Time: 0.5 day.
- Either obtain permission from Ride Austin to use a sample of their data, or generate new data by bootstrapping Toronto trips.  Rework test suite to use sample data only.  Time: 2 days.
- Insert the minimum cost flow and mixed-integer programming linking algorithms in as is, with caveats that these codes are untested.
- Create a white paper using [Jupyter Books](https://jupyter.org/jupyter-book/intro).  The white paper should go into the absolute basics of how each linking algorithm works, and summarize the results found in the `bdit_vfh/trip_linking` notebooks.  Time: 14 days, contingent on public release of the VFH report as we will need to refer to eg. the actual distribution of passenger wait times and the relationship between number of trips and number of distinct drivers per hour.
- Create a short Jupyter notebook tutorial on how to use `triplinker` with the same data, including with hyperparameter optimization.  Can be based directly off of the test suite.  Time: 1 day.

Probably should have:
- Finish productionization of the minimum cost flow and mixed-integer programming linking solutions.  Time: 1 day.
- Finish the simple driver agent modelling to batched linking, possibly refactoring the batched linker in the process.  Time: 3 days.
- Work with Evan and Raph to revise the class inheritance structure of the code, then refactor and modify test suite correspondingly.  Related to the simple driver agent modelling.  Time: 2 days.

Nice to have:
- Full testing and discussion of the MCF/MIP algorithms and simple driver agent models, in particular whether they can solve the discrepancy between linked and empirical fleet size and the significant underestimate of cruising time.

Completely out there:
- Generate a feasible links graph of March 15th, 2017 for the Ride Austin data, then try a hyperparameter optimization (with the full MCF/MIP/agent model) to see if we can jointly match the passenger *and* driver wait times.  This would require we acquire travel time data from Austin.  Raph would then have to upload the Austin centreline and Ride Austin O/D data, and produce a feasibility graph with pgRouting.  Time: 2 weeks for both myself and Raph.

If we complete even the completely out there tasks, we may be in a good position to actually publish our work in a transportation journal.  Alternatively, some of these ideas could be pitched to Francisco or others at UTTRI for follow-up.
